### ISSUE TITLE
Insert not affect / jdbc batch not flushed with ebean-spring-txn and Spring @Transactional

### ISSUE THREAD

**Issue by spinachomes** [ebean-orm/ebean#2086]
insert not affect when transactional method invoke async method in springboot

1、config.setExternalTransactionManager(new SpringJdbcTransactionManager());

2、
@org.springframework.transaction.annotation.Transactional
    public void addCC(AddApplyCCForm form) {
        Set<String> ccEmpIds = form.getCcEmpIds();
        Map<String, AdminUser> userMap = adminUserService.findByEmpId(ccEmpIds);
        AdminUser cuurentUser = ShiroUtil.currentUser();
        for (String ccEmpId : ccEmpIds) {
            ApplyCC cc = new ApplyCC();
            cc.setApplyId(form.getApplyId());
            cc.setReason(form.getReason());
            cc.setFiles(form.getFiles());
            cc.setCcEmpId(ccEmpId);
            cc.setCcName(userMap.get(ccEmpId).getName());
            cc.setRead(false);
            cc.setCreatedUserName(cuurentUser.getName());
            **cc.insert(); // not affect**
        }
        //this is a async method with @Async
        **feishuService.sendCcMessageCard(form.getApplyId(), ccEmpIds, cuurentUser.getName());**
    }

**Comment by rbygrave** [comment]
> not affect

"Not affect" isn't clear enough.  You need to state more clearly what is happening and what isn't happening in terms of sql executed and transaction begin/commit etc.

You should include logging that includes `io.ebean.SQL` and `io.ebean.TXN` which then makes it clearer for anyone trying to help what is actually occurring.  You can then outline and compare the differences that you have looked at (Spring @Transactional vs Ebean @Transactional   and sync vs async etc).

**Comment by spinachomes** [comment]
2020-10-21 08:51:56.002 DEBUG 31595 --- [nio-8080-exec-1] io.ebean.SQL                             : txn[s370202439] select t0.id, t0.name, t0.emp_id, t0.charge_id, t0.leader_id, t0.department_id, t0.department_name, t0.phone, t0.email, t0.password, t0.url72, t0.version from oa_user t0 where t0.emp_id = any(?); --bind(Array[1]={32d73df7})
2020-10-21 08:51:56.002 DEBUG 31595 --- [nio-8080-exec-1] io.ebean.SUM                             : txn[s370202439] FindMany type[AdminUser] origin[DM-Ob1.A.A] exeMicros[11736] rows[1] predicates[t0.emp_id = any(?)] bind[Array[1]={32d73df7}]
2020-10-21 08:51:56.012  INFO 31595 --- [nio-8080-exec-1] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService
2020-10-21 08:51:56.016 DEBUG 31595 --- [nio-8080-exec-1] io.ebean.TXN                             : txn[s370202439] Commit
2020-10-21 08:51:56.040 DEBUG 31595 --- [        Async-1] io.ebean.SQL                             : txn[1002] select t0.id, t0.apply_num, t0.created_user_name, t0.created_department_id, t0.created_department_name, t0.process_definition_key, t0.process_instance_id, t0.status, t0.status_desc, t0.apply_detail, t0.version, t0.created_at, t0.created_by from oa_apply_info t0 where t0.id = ?  ; --bind(1, )
2020-10-21 08:51:56.042 DEBUG 31595 --- [        Async-1] io.ebean.SUM                             : txn[1002] FindBean type[ApplyInfo] origin[1uKuc.A.A] exeMicros[3327] rows[1] bind[1, ]
2020-10-21 08:51:56.044 DEBUG 31595 --- [        Async-1] io.ebean.SQL                             : txn[1003] select t0.id, t0.name, t0.emp_id, t0.charge_id, t0.leader_id, t0.department_id, t0.department_name, t0.phone, t0.email, t0.password, t0.url72, t0.version from oa_user t0 where t0.id = ?  ; --bind(13, )
2020-10-21 08:51:56.044 DEBUG 31595 --- [        Async-1] io.ebean.SUM                             : txn[1003] FindBean type[AdminUser] origin[DM-Oby.A.A] exeMicros[653] rows[1] bind[13, ]
2020-10-21 08:51:56.080 DEBUG 31595 --- [        Async-1] io.ebean.SQL                             : txn[1004] select t0.id, t0.process_key, t0.emp_ids, t0.detail_path, t0.detail_pc_path, t0.version from oa_process_setting t0 where t0.process_key = ?; --bind(process_of_vehicle_use)
2020-10-21 08:51:56.080 DEBUG 31595 --- [        Async-1] io.ebean.SUM                             : txn[1004] FindMany type[ProcessSetting] origin[Byps9G.A.A] exeMicros[1164] rows[1] predicates[t0.process_key = ?] bind[process_of_vehicle_use]
2020-10-21 08:51:56.466 DEBUG 31595 --- [        Async-1] n.yetech.workflow.feishu.FeishuService   : {"code":0,"data":{"message_id":"om_7337ee085241dea400ad5aacaf1f926d"},"msg":"ok"}

**Comment by spinachomes** [comment]
sendCcMessageCard  change to sync,log:
2020-10-21 08:48:56.959 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SQL                             : txn[s1669855231] select t0.id, t0.name, t0.emp_id, t0.charge_id, t0.leader_id, t0.department_id, t0.department_name, t0.phone, t0.email, t0.password, t0.url72, t0.version from oa_user t0 where t0.emp_id = any(?); --bind(Array[1]={32d73df7})
2020-10-21 08:48:56.959 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SUM                             : txn[s1669855231] FindMany type[AdminUser] origin[DM-Ob1.A.A] exeMicros[15574] rows[1] predicates[t0.emp_id = any(?)] bind[Array[1]={32d73df7}]
2020-10-21 08:49:00.628 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SUM                             : txn[s1669855231] BatchControl flush [ApplyCC:0 i:1]
2020-10-21 08:49:00.635 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SQL                             : txn[s1669855231] insert into oa_apply_cc (apply_id, reason, files, cc_emp_id, cc_name, read, created_user_name, version, created_date, created_by) values (?,?,?,?,?,?,?,?,?,?)
2020-10-21 08:49:00.635 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SQL                             : txn[s1669855231]  -- bind(1,昆明,[AttachDto(id=1, fileName=fileName, size=大小)],32d73df7,李清波,false,李清波,1,2020-10-21T08:49:00.579,13)
2020-10-21 08:49:00.650 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SUM                             : txn[s1669855231] Inserted [ApplyCC] [4]
2020-10-21 08:49:00.653 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SQL                             : txn[s1669855231] select t0.id, t0.apply_num, t0.created_user_name, t0.created_department_id, t0.created_department_name, t0.process_definition_key, t0.process_instance_id, t0.status, t0.status_desc, t0.apply_detail, t0.version, t0.created_at, t0.created_by from oa_apply_info t0 where t0.id = ?  ; --bind(1, )
2020-10-21 08:49:00.655 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SUM                             : txn[s1669855231] FindBean type[ApplyInfo] origin[1uKuc.A.A] exeMicros[3080] rows[1] bind[1, ]
2020-10-21 08:49:00.689 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SQL                             : txn[s1669855231] select t0.id, t0.process_key, t0.emp_ids, t0.detail_path, t0.detail_pc_path, t0.version from oa_process_setting t0 where t0.process_key = ?; --bind(process_of_vehicle_use)
2020-10-21 08:49:00.689 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.SUM                             : txn[s1669855231] FindMany type[ProcessSetting] origin[Byps9G.A.A] exeMicros[1184] rows[1] predicates[t0.process_key = ?] bind[process_of_vehicle_use]
2020-10-21 08:49:01.127 DEBUG 30905 --- [nio-8080-exec-3] n.yetech.workflow.feishu.FeishuService   : {"code":0,"data":{"message_id":"om_ba902cff6e36dea93e95bec6a71bedb1"},"msg":"ok"}
2020-10-21 08:49:01.127 DEBUG 30905 --- [nio-8080-exec-3] io.ebean.TXN                             : txn[s1669855231] Commit

**Comment by spinachomes** [comment]
without Transactional and sendCcMessageCard is async:
2020-10-21 08:55:43.308 DEBUG 32173 --- [nio-8080-exec-1] io.ebean.SQL                             : txn[1002] select t0.id, t0.name, t0.emp_id, t0.charge_id, t0.leader_id, t0.department_id, t0.department_name, t0.phone, t0.email, t0.password, t0.url72, t0.version from oa_user t0 where t0.emp_id = any(?); --bind(Array[1]={32d73df7})
2020-10-21 08:55:43.308 DEBUG 32173 --- [nio-8080-exec-1] io.ebean.SUM                             : txn[1002] FindMany type[AdminUser] origin[DM-Ob1.A.A] exeMicros[11373] rows[1] predicates[t0.emp_id = any(?)] bind[Array[1]={32d73df7}]
2020-10-21 08:55:43.318 DEBUG 32173 --- [nio-8080-exec-1] io.ebean.SUM                             : txn[1003] BatchControl flush [ApplyCC:0 i:1]
2020-10-21 08:55:43.325 DEBUG 32173 --- [nio-8080-exec-1] io.ebean.SQL                             : txn[1003] insert into oa_apply_cc (apply_id, reason, files, cc_emp_id, cc_name, read, created_user_name, version, created_date, created_by) values (?,?,?,?,?,?,?,?,?,?)
2020-10-21 08:55:43.325 DEBUG 32173 --- [nio-8080-exec-1] io.ebean.SQL                             : txn[1003]  -- bind(1,昆明,[AttachDto(id=1, fileName=fileName, size=大小)],32d73df7,李清波,false,李清波,1,2020-10-21T08:55:43.314,13)
2020-10-21 08:55:43.329 DEBUG 32173 --- [nio-8080-exec-1] io.ebean.SUM                             : txn[1003] Inserted [ApplyCC] [6]
2020-10-21 08:55:43.331 DEBUG 32173 --- [nio-8080-exec-1] io.ebean.TXN                             : txn[1003] Commit
2020-10-21 08:55:43.333  INFO 32173 --- [nio-8080-exec-1] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService
2020-10-21 08:55:43.359 DEBUG 32173 --- [        Async-1] io.ebean.SQL                             : txn[1004] select t0.id, t0.apply_num, t0.created_user_name, t0.created_department_id, t0.created_department_name, t0.process_definition_key, t0.process_instance_id, t0.status, t0.status_desc, t0.apply_detail, t0.version, t0.created_at, t0.created_by from oa_apply_info t0 where t0.id = ?  ; --bind(1, )
2020-10-21 08:55:43.361 DEBUG 32173 --- [        Async-1] io.ebean.SUM                             : txn[1004] FindBean type[ApplyInfo] origin[1uKuc.A.A] exeMicros[2926] rows[1] bind[1, ]
2020-10-21 08:55:43.363 DEBUG 32173 --- [        Async-1] io.ebean.SQL                             : txn[1005] select t0.id, t0.name, t0.emp_id, t0.charge_id, t0.leader_id, t0.department_id, t0.department_name, t0.phone, t0.email, t0.password, t0.url72, t0.version from oa_user t0 where t0.id = ?  ; --bind(13, )
2020-10-21 08:55:43.363 DEBUG 32173 --- [        Async-1] io.ebean.SUM                             : txn[1005] FindBean type[AdminUser] origin[DM-Oby.A.A] exeMicros[624] rows[1] bind[13, ]
2020-10-21 08:55:43.395 DEBUG 32173 --- [        Async-1] io.ebean.SQL                             : txn[1006] select t0.id, t0.process_key, t0.emp_ids, t0.detail_path, t0.detail_pc_path, t0.version from oa_process_setting t0 where t0.process_key = ?; --bind(process_of_vehicle_use)
2020-10-21 08:55:43.395 DEBUG 32173 --- [        Async-1] io.ebean.SUM                             : txn[1006] FindMany type[ProcessSetting] origin[Byps9G.A.A] exeMicros[1062] rows[1] predicates[t0.process_key = ?] bind[process_of_vehicle_use]
2020-10-21 08:55:43.768 DEBUG 32173 --- [        Async-1] n.yetech.workflow.feishu.FeishuService   : {"code":0,"data":{"message_id":"om_9b44497622d1f3d497cf2c8554e7eeff"},"msg":"ok"}

**Comment by spinachomes** [comment]
delete config.setExternalTransactionManager and with ebean Transactional and sendCcMessageCard is async:
2020-10-21 09:00:02.055 TRACE 32976 --- [nio-8080-exec-1] io.ebean.TXN                             : txn[1002] Begin
2020-10-21 09:00:02.087 DEBUG 32976 --- [nio-8080-exec-1] io.ebean.SQL                             : txn[1002] select t0.id, t0.name, t0.emp_id, t0.charge_id, t0.leader_id, t0.department_id, t0.department_name, t0.phone, t0.email, t0.password, t0.url72, t0.version from oa_user t0 where t0.emp_id = any(?); --bind(Array[1]={32d73df7})
2020-10-21 09:00:02.088 DEBUG 32976 --- [nio-8080-exec-1] io.ebean.SUM                             : txn[1002] FindMany type[AdminUser] origin[DM-Ob1.A.A] exeMicros[12099] rows[1] predicates[t0.emp_id = any(?)] bind[Array[1]={32d73df7}]
2020-10-21 09:00:02.097  INFO 32976 --- [nio-8080-exec-1] o.s.s.concurrent.ThreadPoolTaskExecutor  : Initializing ExecutorService
2020-10-21 09:00:02.099 DEBUG 32976 --- [nio-8080-exec-1] io.ebean.SUM                             : txn[1002] BatchControl flush [ApplyCC:0 i:1]
2020-10-21 09:00:02.107 DEBUG 32976 --- [nio-8080-exec-1] io.ebean.SQL                             : txn[1002] insert into oa_apply_cc (apply_id, reason, files, cc_emp_id, cc_name, read, created_user_name, version, created_date, created_by) values (?,?,?,?,?,?,?,?,?,?)
2020-10-21 09:00:02.107 DEBUG 32976 --- [nio-8080-exec-1] io.ebean.SQL                             : txn[1002]  -- bind(1,昆明,[AttachDto(id=1, fileName=fileName, size=大小)],32d73df7,李清波,false,李清波,1,2020-10-21T09:00:02.093,13)
2020-10-21 09:00:02.111 DEBUG 32976 --- [        Async-1] io.ebean.SQL                             : txn[1003] select t0.id, t0.apply_num, t0.created_user_name, t0.created_department_id, t0.created_department_name, t0.process_definition_key, t0.process_instance_id, t0.status, t0.status_desc, t0.apply_detail, t0.version, t0.created_at, t0.created_by from oa_apply_info t0 where t0.id = ?  ; --bind(1, )
2020-10-21 09:00:02.112 DEBUG 32976 --- [nio-8080-exec-1] io.ebean.SUM                             : txn[1002] Inserted [ApplyCC] [7]
2020-10-21 09:00:02.113 DEBUG 32976 --- [        Async-1] io.ebean.SUM                             : txn[1003] FindBean type[ApplyInfo] origin[1uKuc.A.A] exeMicros[3824] rows[1] bind[1, ]
2020-10-21 09:00:02.114 DEBUG 32976 --- [nio-8080-exec-1] io.ebean.TXN                             : txn[1002] Commit
2020-10-21 09:00:02.116 DEBUG 32976 --- [        Async-1] io.ebean.SQL                             : txn[1004] select t0.id, t0.name, t0.emp_id, t0.charge_id, t0.leader_id, t0.department_id, t0.department_name, t0.phone, t0.email, t0.password, t0.url72, t0.version from oa_user t0 where t0.id = ?  ; --bind(13, )
2020-10-21 09:00:02.117 DEBUG 32976 --- [        Async-1] io.ebean.SUM                             : txn[1004] FindBean type[AdminUser] origin[DM-Oby.A.A] exeMicros[2065] rows[1] bind[13, ]
2020-10-21 09:00:02.136 DEBUG 32976 --- [        Async-1] io.ebean.SQL                             : txn[1005] select t0.id, t0.process_key, t0.emp_ids, t0.detail_path, t0.detail_pc_path, t0.version from oa_process_setting t0 where t0.process_key = ?; --bind(process_of_vehicle_use)
2020-10-21 09:00:02.137 DEBUG 32976 --- [        Async-1] io.ebean.SUM                             : txn[1005] FindMany type[ProcessSetting] origin[Byps9G.A.A] exeMicros[1886] rows[1] predicates[t0.process_key = ?] bind[process_of_vehicle_use]
2020-10-21 09:00:02.562 DEBUG 32976 --- [        Async-1] n.yetech.workflow.feishu.FeishuService   : {"code":0,"data":{"message_id":"om_002eb2a5be68f2bc3c4032f61d0bf1ca"},"msg":"ok"}

**Comment by spinachomes** [comment]
with ebean Transactional insert successfully

**Comment by spinachomes** [comment]
@rbygrave  I find this issue has nothing with @aysc method,if parent method use spring Transactional,call getId after insert() ,record insert sucessfully

**Comment by rbygrave** [comment]
Ok.  Do you want to post up another version of the code / test case that shows the behaviour?  (With the associated logs)

Thanks

**Comment by spinachomes** [comment]
@Service
@Scope(proxyMode = ScopedProxyMode.TARGET_CLASS)
public class TestEbeanService {

    @org.springframework.transaction.annotation.Transactional
    public void testInsert() {
        ApplyInfo info = new ApplyInfo();
        info.insert();
    }
}

@SpringBootTest
@ActiveProfiles("dev")
public class TestEbean {

    @Resource
    private TestEbeanService testEbeanService;

    @Test
    void test() {
        testEbeanService.testInsert();
    }
}

**Comment by spinachomes** [comment]
does not call info.getId() ebean log:
2020-10-30 08:49:47.401 DEBUG 70137 --- [           main] io.ebean.TXN                             : txn[s2046545652] Commit

**Comment by rbygrave** [comment]
>  call `info.getId()` 

So I'm thinking configuration wise jdbc batch has been turned on (probably via persistBatch) ?

So if that is the case then the insert is being done using batch mode (fine).  With the call to `info.getId()` that is triggering a flush of the batch (hence we see the insert statements execute at that point in the logs).

Normally the batch is also flushed on transaction commit ... so it seems that Spring `@Transactional` is not triggering a flush.  

Can you prepare an example application that reproduces this issue and then I can clone it and run it.

Thanks, Rob.

**Comment by spinachomes** [comment]
yeah,I set persistBatching to true and batch size is 100.
the full demo:[https://github.com/spinachomes/demo.git](url),please find DemoDomainServiceTest.java to run tests

**Comment by rbygrave** [comment]
So the fix for this issue is in ebean-spring-txn version 12.4.2.

The effective fix was to SpringTxnListener to perform `transaction.flushBatch();` on `beforeCommit()` 

```java
    @Override
    public void beforeCommit(boolean readOnly) {
      if (!readOnly) {
        transaction.flushBatch();
      }
    }
```

**Comment by rbygrave** [comment]
Fixed in ebean-spring-txn version 12.4.2
